src/utils.c

```diff
@@ -1025,6 +1025,270 @@ ts_relation_size_impl(Oid relid)
 	return relsize;
 }
 
+/*
+ * Try to get cached size for a provided relation across all forks. The
+ * size is returned in terms of number of blocks.
+ *
+ * The function calls the underlying smgrnblocks if there is no cached
+ * data. That call populates the cache for subsequent invocations. This
+ * cached data gets removed asynchronously by PG relcache invalidations
+ * and then the refresh/cache cycle repeats till the next invalidation.
+ */
+static int64
+ts_try_relation_cached_size(Relation rel, bool verbose)
+{
+	BlockNumber result = 0, nblocks = 0;
+	ForkNumber forkNum;
+	bool cached = true;
+
+	/* Get heap size, including FSM and VM */
+	for (forkNum = 0; forkNum <= MAX_FORKNUM; forkNum++)
+	{
+#if PG14_LT
+		/* PG13 does not have smgr_cached_nblocks */
+		result = InvalidBlockNumber;
+#else
+		result = RelationGetSmgr(rel)->smgr_cached_nblocks[forkNum];
+#endif
+
+		if (result != InvalidBlockNumber)
+		{
+			nblocks += result;
+		}
+		else
+		{
+			if (smgrexists(RelationGetSmgr(rel), forkNum))
```

Comment:
Nit: You might want to move this out of the loop.

Reply:
I don't think we can move it out, right? We check for cached value and only if not and only if the fork exists do we check for its size.

Reply:
Right, of course.